com.reprezen.swagedit/src/com/reprezen/swagedit/validation/Validator.java

```diff
@@ -56,21 +66,85 @@
 		}
 
 		if (jsonContent == null) {
-			return Collections.singleton(new SwaggerError(IMarker.SEVERITY_ERROR, 
-					"Unable to read content. It may be invalid YAML"));
+			errors.add(new SwaggerError(
+					IMarker.SEVERITY_ERROR, 
+					Messages.error_cannot_read_content));
 		} else {
 			final Node yaml = document.getYaml();
 			final ErrorProcessor processor = new ErrorProcessor(yaml);
 
-			ProcessingReport report = null;
-			try {
-				report = schemaManager.getSwaggerSchema().getSchema().validate(jsonContent, true);
-			} catch (ProcessingException e) {
-				return processor.processMessage(e.getProcessingMessage());
+			errors.addAll(validateAgainstSchema(processor, jsonContent));
+			errors.addAll(checkDuplicateKeys(yaml));
+		}
+
+		return errors;
+	}
+
+	/*
+	 * Validates the yaml document against the Swagger schema
+	 */
+	protected Set<SwaggerError> validateAgainstSchema(ErrorProcessor processor, JsonNode jsonContent) {
+		final Set<SwaggerError> errors = Sets.newHashSet();
+
+		try {
+			ProcessingReport report = 
+					schemaManager.getSwaggerSchema()
+					.getSchema()
+					.validate(jsonContent, true);
+
+			errors.addAll(processor.processReport(report));
+		} catch (ProcessingException e) {
+			errors.addAll(processor.processMessage(e.getProcessingMessage()));
+		}
+
+		return errors;
+	}
+
+	/*
+	 * Finds all duplicate keys in all objects present in the yaml document.
+	 */
+	protected Set<SwaggerError> checkDuplicateKeys(Node node) {
+		Set<SwaggerError> errors = Sets.newHashSet();
+		
+		if (node.getNodeId() == NodeId.mapping) {
+			MappingNode n = (MappingNode) node;
+
+			Map<String, Set<Node>> duplicates = Maps.newHashMap();
+			for (NodeTuple tuple: n.getValue()) {
+				Node keyNode = tuple.getKeyNode();
+
+				if (keyNode.getNodeId() == NodeId.scalar) {
+					String key = ((ScalarNode) keyNode).getValue();
+					if (duplicates.containsKey(key)) {
+						duplicates.get(key).add(keyNode);
+					} else {
+						duplicates.put(key, Sets.newHashSet(keyNode));
+					}
+				}
+
+				errors.addAll(checkDuplicateKeys(tuple.getValueNode()));
+			}
+
+			for (String key: duplicates.keySet()) {
+				Set<Node> values = duplicates.get(key);
+				if (values.size() > 1) {
+					for (Node duplicate: values) {
+						errors.add(new SwaggerError(
+								duplicate.getStartMark().getLine() + 1, 
+								IMarker.SEVERITY_WARNING, 
+								String.format(Messages.error_duplicate_keys, key)));
+					}
+				}
```

Comment:
I am thinking of a burden on performance. We have two passes here:
1. to create a map
2. to check that every key has a single value.
@ghillairet , let's think how we can improve it. Maybe, use a map that would throw an exception if you add a value for the same key twice and catch an exception.